### PR TITLE
Add configuration and features to logging

### LINKED ISSUE #1998: Logging configuration to enable/disable specific logs
Subtask of #1840 

Here are different versions I have in mind to enable/disable different logs:

```yaml
telemetry:
  logging:
    format: pretty # Could be json or pretty
    router:
      request:
        enabled: true
        level: trace # Display with trace level (default: debug)
      response:
        on_graphql_error: true # Display a log when we return a GraphQL error
        on_http_error: true # Display a log if status code >200
        always: false # If enabled we log every requests
        level: info # Display with info level (default: debug)

    subgraphs:
      all:
        request: 
          enabled: true
          level: debug # Display with debug level (default: debug)
        response: 
          on_graphql_error: true # Display a log when we return a GraphQL error
          on_http_error: true # Display a log if status code >200
          always: false # If enabled we log every requests
      my_subgraph:
        request: 
          enabled: true
        response:
          on_graphql_error: true # Display a log when we return a GraphQL error
          on_http_error: true # Display a log if status code >200
          level: info # Display with info level (default: debug)
          always: false # If enabled we log every requests
``` 

or with a mode:

```yaml
telemetry:
  logging:
    format: pretty # Could be json or pretty
    router:
      request:
        enabled: true
        level: info # log the request on info level
      response:
        # on_error includes http_error and graphql_error 
        mode: on_error # on_error | always
        level: debug # log the response on debug level
    subgraphs:
      all:
        request:
          enabled: true
          level: trace # log the response on trace level
        response: 
          # on_error includes http_error and graphql_error 
          mode: on_error # on_error | always
      my_subgraph:
        request: 
          enabled: true
        response:
          # on_error includes http_error and graphql_error 
          mode: on_error # on_error | always
```

### PR DESCRIPTION
**Pull request by bnjjj**
close #1998

### CODE REVIEW

File: apollo-router/Cargo.toml [hunk 1]
```diff
@@ -184,11 +184,11 @@ tower-service = "0.3.2"
 # https://github.com/apollographql/router/pull/1509.  A comment which exists
 # there (and above around the `opentelemetry` packages) should this ever change.
 # There is also a note below on the `^tracing` dev-dependencies!
-tracing = "=0.1.34"
-tracing-core = "=0.1.26"
+tracing = "=0.1.35"
```
Comment (Geal) [inline apollo-router/Cargo.toml:187]: should we remove that condition on the tracing version now that the insta failures appear resolved?

File: apollo-router/src/plugins/telemetry/config.rs [hunk 2]
```diff
@@ -86,6 +89,130 @@ pub(crate) struct Tracing {
...
+pub(crate) const fn default_display_filename() -> bool {
+    true
+}
+
+pub(crate) const fn default_display_line_number() -> bool {
+    true
+}
+
+impl Logging {
+    pub(crate) fn get_attributes_to_exclude(&self) -> AttributesToExclude {
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/config.rs:117]: the logic is a bit hard to follow here, could you add some comments explaining what is suppsoed to happen? is `SPECIFIC_ATTRIBUTES` the list of default attributes?

File: apollo-router/src/plugins/telemetry/filtering_layer.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+{
+    fn enabled(
+        &self,
+        metadata: &tracing::Metadata<'_>,
+        ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        // The logic is splitted between `enabled` and `event_enabled` because we can't evaluate the value of an attribute in this method
+        // This method is evaluated only once for a specific callsite, which means if I return false here I won't re-enter in this method everytime for this event
+        // Which is better for performance. So for supergraph level we can use this method, for subgraph level we have to know the value of subgraph attribute
+        if metadata.level() == &Level::DEBUG && metadata.is_event() {
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/filtering_layer.rs:51]: why is attribute filtering only active at the `DEBUG` level?

File: apollo-router/src/plugins/telemetry/filtering_layer.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        if field_name == SUBGRAPH_ATTRIBUTE_NAME {
+                            subgraph = true;
+                            // Cannot do anything for subgraph at this level. It will be in `event_enabled` method.
+                            break;
+                        }
+                        if attributes_to_exclude.supergraph.contains(field_name) {
+                            to_exclude = true;
+                        }
+                    }
+                    if !subgraph && to_exclude {
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/filtering_layer.rs:66]: so there's something that's not clear to me here: is the goal of that feature to add/remove some attributes from logs, or to remove logs containing some specific attributes? If it's the latter, could we rename the  `include_attributes` configuration option to `contains_attributes` or `has_attributes`? Because `include_attributes` feels a bit like "add these attributes to the log messages"

Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:66]: Yes it doesn't show the entire event. Because the tracing API doesn't let us to filter or to hide specific attributes on an event. The way to do that is to use a custom formatter, but if we do this, it means for JSON formatter we will have to copy/paste the code of JSON formatter from tracing codebase and it's a big chunk of code. BUT... in the future... I already opened a PR to add new feature on the tracing API to be able to filter attributes. But it's a more long term solution because you know, waiting for the PR to be merged, and then for the release and so one ....

**Comment by BrynCooke** [comment]
Couple of high level observations:
* It would be nice to not have separate log events for subgraph request and response attributes. It feels like this should be one event for request and one for response.
* I would expect that events should be at info. Debug events shouldn't need config in our yaml and can already be controlled via RUST_LOG or APOLLO_LOG. What'll happen is that users will configure this but the events won't make it to their APM unless they also set their log level to debug.

File: apollo-router/src/plugins/telemetry/filtering_layer.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:51]: Because these logs are always in DEBUG, but following a discussion with Bryn it will be in INFO

File: apollo-router/src/plugins/telemetry/config.rs (hunk 2 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/config.rs:117]: Yes, bad namings. Will add more comments :)

File: apollo-router/Cargo.toml (hunk 1 above)
Comment (bnjjj) [reply apollo-router/Cargo.toml:187]: That would make sense yes

File: apollo-router/src/plugins/telemetry/filtering_layer.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+use std::collections::HashSet;
+
+use tracing::field;
+use tracing::Level;
+use tracing::Subscriber;
+use tracing_core::Field;
+use tracing_subscriber::Layer;
+
+// Specific attributes for logging
+pub(crate) const SPECIFIC_ATTRIBUTES: [&str; 4] = [
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/filtering_layer.rs:11]: Seconding @Geal comment about this name "SPECIFIC" seems odd. If you are trying to avoid "DEFAULT", then perhaps consider "USUAL" or "REQUIRED" (if they are required).

File: apollo-router/src/plugins/telemetry/filtering_layer.rs (hunk 3 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:51]: Agree that seems strange. Should be mentioned in the documentation, but I don't really understand why it exists. Also, if DEBUG is special for some reason, don't we also want to process when level is TRACE?

File: apollo-router/src/plugins/telemetry/filtering_layer.rs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                if let Some(subgraph) = fields_visitor.fields.get(SUBGRAPH_ATTRIBUTE_NAME) {
+                    if let Some(attrs_to_exclude) = attributes_to_exclude
+                        .subgraphs
+                        .get(subgraph)
+                        .map(|a| a.iter().map(|a| a.as_str()).collect::<HashSet<&str>>())
+                    {
+                        if !attrs_to_exclude.is_disjoint(&field_names) {
+                            return false;
+                        }
+                    } else if !attributes_to_exclude
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/filtering_layer.rs:106]: Implementation consideration: Why maintain `all_subgraphs` as a backstop? Why not de-normalise the `all_subgraphs` values into each individual `subgraph`? May take slightly more memory, but would simplify and possibly speed up processing.

**Review by garypen — COMMENTED** [review]
A lot of this looks great. I've got one extra question: What does the performance data look like for this change? And, not related to this change, but something I think about now and then, does our perf testing actually do any logging?

File: apollo-router/src/plugins/telemetry/filtering_layer.rs (hunk 5 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:11]: Go for `DEFAULT` ! Thanks :)

**Comment by bnjjj** [comment]
@gary We have logs enabled in our performance test. So after launching this on performance test it has no impact on the performance. BUT... the logging configuration is not set at all. One thing to keep in mind and also that was the main reason for me to write a tracing Layer is for example when it enters in the `enabled` method it will be called only one time for each callsite, it means only one time for each log `tracing::info/debug/...` in our code. If this method returns `false` it won't go back again in this method because this callsite has been disabled so it won't do the process again and again for each call. It's a little bit different for the `on_event_enabled` method because in this method we're looking at the value of the subgraph attribute.

File: apollo-router/src/plugins/telemetry/filtering_layer.rs (hunk 6 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:106]: Hm I'm not sure it would speed up processing, maybe simplify yes. But it means we need to access to global configuration or global parsed supergraph to know what's the list of subgraphs. I don't think we have a clean implementation of this currently but that might be something we should provide in the `PluginInit` struct ?

Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:106]: I deleted the `DEBUG` level condition in the process

File: apollo-router/src/plugins/telemetry/filtering_layer.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/filtering_layer.rs:51]: I deleted the DEBUG level condition in the process

**Comment by bnjjj** [comment]
New configuration suggestion:

```yaml
# Log every request, but this is never on by default.
logging:
  requests: true

# Explicitly on in --dev, documented as such.
# Also possible to just set it and have it on in production
logging:
  requests:
     with_header_present: apollo-router-log-request

## MAYBE

logging:
  requests:
     when_header_present: apollo-router-log-request
     request_headers: true
     request_body: true
     request_body: true
```

**Comment by bnjjj** [comment]
New suggestion https://github.com/apollographql/router/issues/1998#issuecomment-1318741010

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 7]
```diff
@@ -428,42 +446,64 @@ impl Telemetry {
...
-                        .map_fmt_fields(|_f| JsonFields::new())
-                        .finish()
-                        .with(telemetry);
-                    if let Err(e) = set_global_default(subscriber) {
-                        ::tracing::error!("cannot set global subscriber: {:?}", e);
+                    match config
+                        .logging
+                        .as_ref()
+                        .map(|l| l.format)
+                        .unwrap_or_default()
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/mod.rs:475]: Will this result in a change in behaviour? I'm not sure what the default format is, but I wonder if that will match up with the existing tty detecting behaviour that we have.
If it is a change, we need to discuss that and maybe mark it as breaking change in the changelog.

File: docs/source/configuration/logging.mdx [hunk 8]
```diff
@@ -66,6 +66,35 @@ JSON-formatted logging provides compatibility with common searchable logging too
...
 ```
 
+## Basic configuration
+
+### Disclaimer
+
+This is part of an experimental feature, it means any time until it's stabilized (without the prefix `experimental_`) we could change the configuration shape or adding/removing features.
+If you want to give feedback or participate in that feature feel free to join [this discussion on GitHub](https://github.com/apollographql/router/discussions/1961).
+
+By default some of our logs containing sensible data (like request body, response body, headers) are not displayed even if we are in the right log level.
```
Comment (garypen) [inline docs/source/configuration/logging.mdx:76]: I'm not familiar with the usage of "sensible" in this context. Do you mean "sensitive"?

**Review by garypen — APPROVED** [review]
Approving, but I'd like to understand if there's a change in behaviour before merging.

File: apollo-router/src/plugins/telemetry/mod.rs (hunk 7 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/mod.rs:475]: It won't change the default behavior thanks to this impl https://github.com/apollographql/router/pull/2040/files/6dc13330f5e96e475ad3f0b748a0201de2afad69#diff-eb7847da8c5109309df2eacc8f7acd6c73d93260c81890258b37b9dac3f30f7cR239

File: docs/source/configuration/logging.mdx (hunk 8 above)
Comment (bnjjj) [reply docs/source/configuration/logging.mdx:76]: Oups yes. Sorry :)
